NIFI-1002: Added WebSocket support.

ijokarumawak opened the pull request:
Thank you for submitting a contribution to Apache NiFi.

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [x] Is there a JIRA ticket associated with this PR? Is it referenced 
     in the commit message?

- [x] Does your PR title start with NIFI-XXXX where XXXX is the JIRA number you are trying to resolve? Pay particular attention to the hyphen "-" character.

- [x] Has your PR been rebased against the latest commit within the target branch (typically master)?

- [x] Is your initial contribution a single, squashed commit?

### For code changes:
- [x] Have you ensured that the full suite of tests is executed via mvn -Pcontrib-check clean install at the root nifi folder?
- [x] Have you written or updated unit tests to verify your changes?
- [x] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)? 
- [x] If applicable, have you updated the LICENSE file, including the main LICENSE file under nifi-assembly?
- [x] If applicable, have you updated the NOTICE file, including the main NOTICE file found under nifi-assembly?
- [ ] If adding new Properties, have you added .displayName in addition to .name (programmatic access) for each of the new properties?

### For documentation related changes:
- [x] Have you ensured that format looks appropriate for the output in which it is rendered?

### Note:
Please ensure that once the PR is submitted, you check travis-ci for build issues and submit an update to your PR as soon as possible.

ijokarumawak commented:
Hello NiFi developers,

This PR contains a new set of WebSocket support components.

At least there're two WebSocket processors available on GitHub already, [nifi-websocket](https://github.com/xmlking/nifi-websocket) by @xmlking , and [nifi-websockets-bundle](https://github.com/acesir/nifi-websockets-bundle) by @acesir. I appreciate these projects providing WebSocket support for NiFi, and wanted to merge the capability to Apache NiFi code base. I referred these existing projects, and add few functionalities and Unit Test codes.

Features:
- As a WebSocket client to connect a remote WebSocket server
- As a WebSocket server to receive connection from remote WebSocket clients
- Receive and send text messages as well as binary messages
- Secure communication (using SslContextService) is supported

It may look more complex, so for the description purpose, I wrote a blog post about how it works, and how it's designed. Please check this post as well, [NiFi WebSocket Support](http://ijokarumawak.github.io/nifi/2016/11/04/nifi-websocket/).

Any comment or advice would be appreciated, thanks in advance!

olegz commented on nifi-nar-bundles/nifi-websocket-bundle/nifi-websocket-processors/src/main/java/org/apache/nifi/processors/websocket/AbstractWebSocketProcessor.java:
```diff
@@ -0,0 +1,80 @@
+import java.util.ArrayList;
+import java.util.List;
+
+public abstract class AbstractWebSocketProcessor extends AbstractSessionFactoryProcessor {
```
Is there a reason why this class extends from _ AbstractSessionFactoryProcessor_ and not _AbstractProcessor_? Just trying to figure out the need for _ onTrigger()_ implementation that is identical to the one in __AbstractProcessor_.

ijokarumawak replied:
@olegz The reason why it extends AbstractSessionFactoryProcessor is to capture `processSessionFactory instance. So that it can create a session at [onStopped()](https://github.com/apache/nifi/pull/1184/files#diff-3347cb9c958c84660d5db88e46808fa2R222) to flush incoming WebSocket messages those only exist on memory. I followed the pattern of AbstractMQTTProcessor.

olegz commented on nifi-nar-bundles/nifi-websocket-bundle/nifi-websocket-processors/src/main/java/org/apache/nifi/processors/websocket/AbstractWebSocketProcessor.java:
```diff
@@ -0,0 +1,80 @@
+
+    static List<PropertyDescriptor> getAbstractPropertyDescriptors(){
+        final List<PropertyDescriptor> descriptors = new ArrayList<>();
+        return descriptors;
```
I am not sure I understand this code
1. For each call you create a new List
2. It's always empty
Could you please clarify?

ijokarumawak replied:
@olegz As you suspected, it's not required, and it's there simply because I ported the basic code structure from MQTT processors. This method was used for sharing common property descriptors. However, since there's no such properties for now, it can be removed. I will remove this.

olegz commented on nifi-nar-bundles/nifi-websocket-bundle/nifi-websocket-processors/src/main/java/org/apache/nifi/processors/websocket/AbstractWebSocketProcessor.java:
```diff
@@ -0,0 +1,80 @@
+    }
+
+    protected ComponentLog logger;
+    protected ProcessSessionFactory processSessionFactory;
```
Since there is a potential for the above variables to be accessed by different thread, there are thread visibility concerns. Consider making them _volatile_.

ijokarumawak replied:
Added volatile keyword to these variables.

ijokarumawak commented:
@olegz Thanks for taking time to review! I added a commit to address your feedback. Also, I've added `displayName` to the property descriptors those were added in this PR.

olegz commented on nifi-nar-bundles/nifi-websocket-bundle/nifi-websocket-services-jetty/src/main/java/org/apache/nifi/websocket/jetty/JettyWebSocketServer.java:
```diff
@@ -0,0 +1,258 @@
+        servletHandler.addServletWithMapping(JettyWebSocketServlet.class, "/*");
+
+        // Need to specify classloader, otherwise since the callstack doesn't have any nifi specific class, so it can't use nar.
+        try (NarCloseable closeable = NarCloseable.withComponentNarLoader(WebSocketServerFactory.class)) {
```
This no longer compiles after recent work on dynamic class loading by @bbende

ijokarumawak replied:
Thanks Oleg, I removed the use of NarCloseable and confirmed that JettyWebSocketServer works without this now.

olegz commented on nifi-nar-bundles/nifi-websocket-bundle/nifi-websocket-processors/src/main/java/org/apache/nifi/processors/websocket/AbstractWebSocketGatewayProcessor.java:
```diff
@@ -0,0 +1,314 @@
+            .defaultValue("10000")
+            .build();
+
+    private volatile LinkedBlockingQueue<WebSocketMessage> incomingMessageQueue;
```
This particular issue keeps coming up. Basically the pattern of maintaining an internal queue for messages may lead to OOM exception as well as lost data if processor crashes while the queue still has data. This is due to the fact that you are essentially integrating with external scheduling mechanism
One way of mitigating this is to rely on such mechanism for distributing messages while essentially ignoring our schedule. You can look at the example of it in SMTP processor https://github.com/apache/nifi/blob/d838f61291d2582592754a37314911b701c6891b/nifi-nar-bundles/nifi-email-bundle/nifi-email-processors/src/main/java/org/apache/nifi/processors/email/ListenSMTP.java#L175
While may not be ideal, it certainly paves a simpler path for future improvement when we get a chance to define an Event-driver Processor abstraction as a first class component of NiFi.

ijokarumawak replied:
Thank you for bringing this up. I had been worrying about this queue, too. To make it simpler, I removed the intermediate queue, and let incoming messages to be transferred to NiFi relationships.

ijokarumawak commented:
@olegz Thanks again for the great review! I rebased the PR and done additional refactoring based on your feedback. Please let me know if you prefer it to be squashed for further review process.

olegz commented:
Koji, this is great and I'll be looking at it shortly with hopes of merging it some time today.

olegz commented:
@ijokarumawak at this point I am going to say Great work! I am sure in month to come there will be things ti improve as people start using it , but LGTM for now. Will merge shortly!

ijokarumawak commented:
@olegz Thank you very much! I hope it will be helpful for users and willing to improve it over time!
